### ISSUE TITLE
CloudFront settings reverting on each deploy

### ISSUE THREAD

**Issue by StephaneBischoff** [aws-amplify/amplify-hosting#2622]
**Please describe which feature you have a question about?**
I am deploying AWS Amplify NextJS app using Git deployment. 
I also apply specific configuration headers manually in AWS Console for the CloudFront distribution. I set "Accept-Language" as a cache key header under Behaviors for Default (*) route in CloudFront .

But, these settings keep reverting. I believe this happens each time I deploy a new version of my site with Amplify.
Is there a way to set this config somewhere in Amplify so that its applied on each deploy? Or maybe even better, make sure Amplify does not update CloudFront?

Region: ca-central

**Comment by hloriana** [comment]
Hi @StephaneBischoff  have you tried adding that header directly in the Amplify Console? Here is a guide for adding the headers: https://docs.aws.amazon.com/amplify/latest/userguide/custom-headers.html#setting-custom-headers

**Comment by yassinecc** [comment]
The TLDR of how I did it:
- Figure out what elements of the Cloudfront configuration should be reverted post deployment with the AWS CLI
- Set up a Lambda function that corrects the Cloudfront configuration with the information from above using the AWS SDK
- Set up an SNS notification that triggers that Lambda function on Amplify deployment

-----

The main idea is to use some AWS SDK commands to programmatically reset the Cloudfront configuration post deployment. First go to the Amplify console > App settings > Notifications and create an email notification, we will need it later. You can delete the email notification afterwards but I recommend keeping it for a while if you need to debug Amplify notifications

A simple way to call the AWS SDK from Lambda functions is to use a NodeJS runtime. You will need to[ setup a specific Lambda layer](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html) to include the AWS JS SDK within the Lambda function's runtime

Once you have created a Lambda function with this layer, add a trigger to your Lambda function and choose the SNS topic that was created when setting up the Amplify email notification. Now your Lambda function will get called on build success / failure

Now you will need to figure out what to reset in the Cloudfront configuration. You can for example locally run `aws cloudfront get-distribution-config --id $DISTRIBUTION_ID` before / after a deployment to see what changes you need to make. You can then use something like the following in your Lambda function:
```js
const { CloudFrontClient, ListCachePoliciesCommand, GetDistributionConfigCommand, UpdateDistributionCommand } = require("@aws-sdk/client-cloudfront");

async function updateCloudfrontDistribution(distributionId) {
    const client = new CloudFrontClient({ region: process.env.AWS_REGION });

    const distributionCommand = new GetDistributionConfigCommand({Id: distributionId});
    const distributionResponse = await client.send(distributionCommand);
    const etag = distributionResponse.ETag;
    const newConfig = { ...distributionResponse.DistributionConfig };
    // Change newConfig as you need
    const updateCommand = new UpdateDistributionCommand({Id: distributionId, DistributionConfig: newConfig, IfMatch: etag});
    await client.send(updateCommand);
}
```

Refer to [the AWS Cloudfront client documentation](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-cloudfront/index.html) for how the commands used above work. Also see [this article](https://betterprogramming.pub/how-to-send-slack-notifications-for-aws-amplify-deployments-99278a778ff7) for screenshots on how to set up the Lambda / Amplify connection

Let me know if it helps, I can provide more details if needed
